pkg/certificate/manager.go

```diff
@@ -152,75 +157,116 @@ func (m *Manager) GetTrustDomain() string {
-				log.Error().Err(err).Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrRotatingCert)).
-					Msgf("Error rotating cert SerialNumber=%s", cert.GetSerialNumber())
-				continue
-			}
 
-			m.msgBroker.GetCertPubSub().Pub(events.PubSubMessage{
-				Kind:   announcements.CertificateRotated,
-				NewObj: newCert,
-				OldObj: cert,
-			}, announcements.CertificateRotated.String())
+	for key, cert := range certs {
+		opts := []IssueOption{}
+		if key == cert.CommonName.String() {
+			opts = append(opts, FullCNProvided())
+		}
 
-			log.Debug().Msgf("Rotated certificate (old SerialNumber=%s) with new SerialNumber=%s", cert.SerialNumber, newCert.SerialNumber)
+		_, err := m.IssueCertificate(key, cert.certType, opts...)
+		if err != nil {
+			log.Error().Err(err).Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrRotatingCert)).
+				Msgf("Error rotating cert SerialNumber=%s", cert.GetSerialNumber())
 		}
 	}
 }
 
+func (m *Manager) getValidityDurationForCertType(ct CertType) time.Duration {
+	switch ct {
+	case Internal:
+		return constants.OSMCertificateValidityPeriod
+	case IngressGateway:
+		return m.ingressCertValidityDuration()
+	case Service:
+		return m.serviceCertValidityDuration()
+	default:
+		log.Debug().Msgf("Unknown certificate type %s provided when getting validity duration", ct)
+		return constants.OSMCertificateValidityPeriod
+	}
+}
+
+// getFromCache returns the certificate with the specified cn from cache if it exists.
+// Note: getFromCache might return an expired or invalid certificate.
 func (m *Manager) getFromCache(key string) *Certificate {
 	certInterface, exists := m.cache.Load(key)
 	if !exists {
 		return nil
 	}
 	cert := certInterface.(*Certificate)
 	log.Trace().Msgf("Certificate found in cache SerialNumber=%s", cert.GetSerialNumber())
-	if cert.ShouldRotate() {
-		log.Trace().Msgf("Certificate found in cache but has expired SerialNumber=%s", cert.GetSerialNumber())
-		return nil
-	}
 	return cert
 }
 
-// IssueCertificate implements Manager and returns a newly issued certificate from the given client.
-func (m *Manager) IssueCertificate(prefix string, opts ...IssueOption) (*Certificate, error) {
-	var err error
-	cert := m.getFromCache(prefix) // Don't call this while holding the lock
+// IssueCertificate returns a newly issued certificate from the given client
+// or an existing valid certificate from the local cache.
+func (m *Manager) IssueCertificate(prefix string, ct CertType, opts ...IssueOption) (*Certificate, error) {
+	cert, err, _ := group.Do(prefix, func() (interface{}, error) {
+		return m.issueCertificate(prefix, ct, opts...)
+	})
+	if err != nil {
+		return nil, err
+	}
+	return cert.(*Certificate), nil
+}
 
-	options := defaultOptions(m.serviceCertValidityDuration)
+func (m *Manager) issueCertificate(prefix string, ct CertType, opts ...IssueOption) (*Certificate, error) {
+	var rotate bool
+	cert := m.getFromCache(prefix) // Don't call this while holding the lock
+	if cert != nil {
```

Comment:
nit: put the nil check in should rotate to reduce nesting

Reply:
I think I prefer the nesting the call to `shouldRotate`. It makes sense to only call `shouldRotate` when the cert is not nil. Happy to make the change if you think moving the nil check to `shouldRotate` improves readability.